InsightsFrontend/src/app/pages/release-graph/release-link.service.ts

```diff
@@ -0,0 +1,99 @@
+import {Injectable} from '@angular/core';
+import {ReleaseNode} from "./release-node.service";
+
+export interface ReleaseLink {
+	id: string;
+	source: string;
+	target: string;
+}
+
+@Injectable({ providedIn: 'root' })
+export class ReleaseLinkService {
+	private static readonly GITHUB_MASTER_BRANCH = 'master';
+
+	createLinks(sortedGroups: Map<string, ReleaseNode[]>[]): ReleaseLink[] {
+		if (!sortedGroups.length) return [];
+
+		const masterBranch = ReleaseLinkService.GITHUB_MASTER_BRANCH;
+		const masterNodes = sortedGroups[0].get(masterBranch) ?? [];
+
+		return [
+			...this.createMasterBranchLinks(masterNodes),
+			...this.createSubBranchLinks(sortedGroups.slice(1), masterNodes)
+		];
+	}
+
+	private createMasterBranchLinks(nodes: ReleaseNode[]): ReleaseLink[] {
+		const links: ReleaseLink[] = [];
+		for (let i = 0; i < nodes.length - 1; i++) {
+			links.push(this.buildLink(nodes[i], nodes[i + 1]));
+		}
+		return links;
+	}
+
+	private createSubBranchLinks(
+			subGroups: Map<string, ReleaseNode[]>[],
+			masterNodes: ReleaseNode[]
+	): ReleaseLink[] {
+		const sortedSubGroups = this.sortSubGroupsByLatestRelease(subGroups);
+		const links: ReleaseLink[] = [];
+
+		for (const subGroup of sortedSubGroups) {
+			const [branch, subNodes] = [...subGroup.entries()][0];
+			if (!subNodes.length) continue;
+
+			const syntheticLink = this.createSyntheticLinkIfExists(branch, subNodes[0], masterNodes);
+			if (syntheticLink) {
+				links.push(syntheticLink);
+			}
+
+			links.push(...this.createSequentialLinks(subNodes));
+		}
+
+		return links;
+	}
+
+	private sortSubGroupsByLatestRelease(groups: Map<string, ReleaseNode[]>[]): Map<string, ReleaseNode[]>[] {
+		return groups.sort((a, b) => {
+			const [, aNodes] = [...a.entries()][0];
+			const [, bNodes] = [...b.entries()][0];
+			return bNodes[0].publishedAt.getTime() - aNodes[0].publishedAt.getTime();
+		});
+	}
+
+	private createSyntheticLinkIfExists(
+			branch: string,
+			firstNode: ReleaseNode,
+			masterNodes: ReleaseNode[]
+	): ReleaseLink | null {
+		const versionMatch = firstNode.label.match(/^v?(\d+)\.(\d+)/);
+		if (!versionMatch) return null;
+
+		const syntheticId = `synthetic-${branch}-${firstNode.id}`;
+		const syntheticNode = masterNodes.find(n => n.id === syntheticId);
+
+		if (!syntheticNode) return null;
+
+		return {
+			id: `${syntheticNode.id}-${firstNode.id}`,
+			source: syntheticNode.id,
+			target: firstNode.id
+		};
+	}
+
+	private createSequentialLinks(nodes: ReleaseNode[]): ReleaseLink[] {
```

Comment:
This method is identical to `createMasterBranchLinks`. That could possibly be cleaned up.